File: tests/test_totals.py
from cashplot.totals import *
from cashplot.transactions import TransactionBalance, transaction_balances, categorize
from cashplot.sources.ing import load_ing_transactions
from cashplot.config import Config

with open("tests/data/test_config.json", "r") as cfp:
    config = Config.loads(cfp.read())
with open("tests/data/test_transactions.csv", "r") as tfp:
    transactions_uncategorized = list(
        map(lambda tr: tr.convert(), load_ing_transactions(tfp)))
    tr_balances = transaction_balances(categorize(
        transactions_uncategorized, config), config)


def test_get_categories():
    assert get_categories(tr_balances) == ['Salary', 'Shopping']


def test_last_year_range():
    ly_range = last_year_range(tr_balances)
    assert ly_range == [datetime.date(2019, 7, 12), datetime.date(2020, 7, 12)]


def test_floor_period():
    assert floor_period(datetime.datetime(2020, 5, 20).date(), Period.YEAR
                       ) == datetime.datetime(2020, 1, 1).date()
    assert floor_period(datetime.datetime(2200, 12, 31).date(), Period.YEAR
                       ) == datetime.datetime(2200, 1, 1).date()
    assert floor_period(datetime.datetime(2000, 1, 1).date(), Period.YEAR
                       ) == datetime.datetime(2000, 1, 1).date()

    assert floor_period(datetime.datetime(2020, 5, 20).date(), Period.QUARTER
                       ) == datetime.datetime(2020, 4, 1).date()
    assert floor_period(datetime.datetime(2200, 12, 31).date(), Period.QUARTER
                       ) == datetime.datetime(2200, 10, 1).date()
    assert floor_period(datetime.datetime(2000, 1, 1).date(), Period.QUARTER
                       ) == datetime.datetime(2000, 1, 1).date()

    assert floor_period(datetime.datetime(2020, 5, 20).date(), Period.MONTH
                       ) == datetime.datetime(2020, 5, 1).date()
    assert floor_period(datetime.datetime(2200, 12, 31).date(), Period.MONTH
                       ) == datetime.datetime(2200, 12, 1).date()
    assert floor_period(datetime.datetime(2000, 1, 1).date(), Period.MONTH
                       ) == datetime.datetime(2000, 1, 1).date()

    assert floor_period(datetime.datetime(2020, 5, 20).date(), Period.WEEK
                       ) == datetime.datetime(2020, 5, 18).date()
    assert floor_period(datetime.datetime(2200, 12, 31).date(), Period.WEEK
                       ) == datetime.datetime(2200, 12, 29).date()
    assert floor_period(datetime.datetime(2000, 1, 1).date(), Period.WEEK
                       ) == datetime.datetime(1999, 12, 27).date()

    assert floor_period(datetime.datetime(2020, 5, 20).date(), Period.DAY
                       ) == datetime.datetime(2020, 5, 20).date()
    assert floor_period(datetime.datetime(2200, 12, 31).date(), Period.DAY
                       ) == datetime.datetime(2200, 12, 31).date()
    assert floor_period(datetime.datetime(2000, 1, 1).date(), Period.DAY
                       ) == datetime.datetime(2000, 1, 1).date()


def test_next_period():
    assert next_period(datetime.datetime(2020, 5, 1).date(), Period.YEAR
                      ) == datetime.datetime(2021, 1, 1).date()
    assert next_period(datetime.datetime(2200, 12, 31).date(), Period.YEAR
                      ) == datetime.datetime(2201, 1, 1).date()
    assert next_period(datetime.datetime(2000, 1, 1).date(), Period.YEAR
                      ) == datetime.datetime(2001, 1, 1).date()

    assert next_period(datetime.datetime(2020, 5, 1).date(), Period.QUARTER
                      ) == datetime.datetime(2020, 7, 1).date()
    assert next_period(datetime.datetime(2200, 12, 31).date(), Period.QUARTER
                      ) == datetime.datetime(2201, 1, 1).date()
    assert next_period(datetime.datetime(2000, 1, 1).date(), Period.QUARTER
                      ) == datetime.datetime(2000, 4, 1).date()

    assert next_period(datetime.datetime(2020, 5, 1).date(), Period.MONTH
                      ) == datetime.datetime(2020, 6, 1).date()
    assert next_period(datetime.datetime(2200, 12, 31).date(), Period.MONTH
                      ) == datetime.datetime(2201, 1, 1).date()
    assert next_period(datetime.datetime(2000, 1, 1).date(), Period.MONTH
                      ) == datetime.datetime(2000, 2, 1).date()

    assert next_period(datetime.datetime(2020, 5, 1).date(), Period.WEEK
                      ) == datetime.datetime(2020, 5, 4).date()
    assert next_period(datetime.datetime(2200, 12, 31).date(), Period.WEEK
                      ) == datetime.datetime(2201, 1, 5).date()
    assert next_period(datetime.datetime(2000, 1, 1).date(), Period.WEEK
                      ) == datetime.datetime(2000, 1, 3).date()

    assert next_period(datetime.datetime(2020, 5, 1).date(), Period.DAY
                      ) == datetime.datetime(2020, 5, 2).date()
    assert next_period(datetime.datetime(2200, 12, 31).date(), Period.DAY
                      ) == datetime.datetime(2201, 1, 1).date()
    assert next_period(datetime.datetime(2000, 1, 1).date(), Period.DAY
                      ) == datetime.datetime(2000, 1, 2).date()


def test_period_thirds():
    assert period_thirds(datetime.datetime(2020, 5, 1).date(), Period.YEAR
                      ) == (datetime.datetime(2020, 5, 2).date(), datetime.datetime(2020, 9, 1).date())
    assert period_thirds(datetime.datetime(2020, 5, 1).date(), Period.QUARTER
                      ) == (datetime.datetime(2020, 5, 1).date(), datetime.datetime(2020, 5, 31).date())
    assert period_thirds(datetime.datetime(2020, 5, 1).date(), Period.MONTH
                      ) == (datetime.datetime(2020, 5, 11).date(), datetime.datetime(2020, 5, 21).date())
    assert period_thirds(datetime.datetime(2020, 5, 1).date(), Period.WEEK
                      ) == (datetime.datetime(2020, 4, 29).date(), datetime.datetime(2020, 5, 1).date())
    assert period_thirds(datetime.datetime(2020, 5, 1).date(), Period.DAY
                      ) == (datetime.datetime(2020, 5, 1, 8, 0), datetime.datetime(2020, 5, 1, 16, 0))

def test_period_halves():
    assert period_halves(datetime.datetime(2020, 5, 1).date(), Period.YEAR
                      ) == datetime.datetime(2020, 7, 2).date()
    assert period_halves(datetime.datetime(2020, 5, 1).date(), Period.QUARTER
                      ) == datetime.datetime(2020, 5, 16).date()
    assert period_halves(datetime.datetime(2020, 5, 1).date(), Period.MONTH
                      ) == datetime.datetime(2020, 5, 16).date()
    assert period_halves(datetime.datetime(2020, 5, 1).date(), Period.WEEK
                      ) == datetime.datetime(2020, 4, 30).date()
    assert period_halves(datetime.datetime(2020, 5, 1).date(), Period.DAY
                      ) == datetime.datetime(2020, 5, 1, 12, 0)


def test_override_thirds():
    assert override_thirds([datetime.datetime(2020, 5, 1).date(), datetime.datetime(2200, 12, 31).date()], Period.YEAR
            ) == ([datetime.datetime(2020, 5, 2).date(), datetime.datetime(2200, 5, 2).date()],
                  [datetime.datetime(2020, 9, 1).date(), datetime.datetime(2200, 9, 1).date()])
    assert override_thirds([datetime.datetime(2020, 5, 1).date(), datetime.datetime(2200, 12, 31).date()], Period.QUARTER
            ) == ([datetime.datetime(2020, 5, 1).date(), datetime.datetime(2200, 10, 31).date()],
                  [datetime.datetime(2020, 5, 31).date(), datetime.datetime(2200, 12, 1).date()])
    assert override_thirds([datetime.datetime(2020, 5, 1).date(), datetime.datetime(2200, 12, 31).date()], Period.MONTH
            ) == ([datetime.datetime(2020, 5, 11).date(), datetime.datetime(2200, 12, 11).date()],
                  [datetime.datetime(2020, 5, 21).date(), datetime.datetime(2200, 12, 21).date()])
    assert override_thirds([datetime.datetime(2020, 5, 1).date(), datetime.datetime(2200, 12, 31).date()], Period.WEEK
            ) == ([datetime.datetime(2020, 4, 29).date(), datetime.datetime(2200, 12, 31).date()],
                  [datetime.datetime(2020, 5, 1).date(), datetime.datetime(2201, 1, 2).date()])
    assert override_thirds([datetime.datetime(2020, 5, 1).date(), datetime.datetime(2200, 12, 31).date()], Period.DAY
            ) == ([datetime.datetime(2020, 5, 1, 8, 0), datetime.datetime(2200, 12, 31, 8, 0)],
                  [datetime.datetime(2020, 5, 1, 16, 0), datetime.datetime(2200, 12, 31, 16, 0)])

def test_override_halves():
    assert override_halves([datetime.datetime(2020, 5, 1).date(), datetime.datetime(2200, 12, 31).date()], Period.YEAR
            ) == [datetime.datetime(2020, 7, 2).date(), datetime.datetime(2200, 7, 2).date()]
    assert override_halves([datetime.datetime(2020, 5, 1).date(), datetime.datetime(2200, 12, 31).date()], Period.QUARTER
            ) == [datetime.datetime(2020, 5, 16).date(), datetime.datetime(2200, 11, 16).date()]
    assert override_halves([datetime.datetime(2020, 5, 1).date(), datetime.datetime(2200, 12, 31).date()], Period.MONTH
            ) == [datetime.datetime(2020, 5, 16).date(), datetime.datetime(2200, 12, 16).date()]
    assert override_halves([datetime.datetime(2020, 5, 1).date(), datetime.datetime(2200, 12, 31).date()], Period.WEEK
            ) == [datetime.datetime(2020, 4, 30).date(), datetime.datetime(2201, 1, 1).date()]
    assert override_halves([datetime.datetime(2020, 5, 1).date(), datetime.datetime(2200, 12, 31).date()], Period.DAY
            ) == [datetime.datetime(2020, 5, 1, 12, 0), datetime.datetime(2200, 12, 31, 12, 0)]


def test_categories_changes():
    year_periods, year_income_changes, year_expenses_changes = categories_changes(tr_balances, Period.YEAR)
    assert year_periods == [datetime.date(2020, 1, 1)]
    assert year_income_changes == {'Salary': [Decimal('2500')],
                              'Shopping': [Decimal('0')]}
    assert year_expenses_changes == {'Salary': [Decimal('0')],
                              'Shopping': [Decimal('50')]}

    quarter_periods, quarter_income_changes, quarter_expenses_changes = categories_changes(tr_balances, Period.QUARTER)
    assert quarter_periods == [datetime.date(2020, 4, 1), datetime.date(2020, 7, 1)]
    assert quarter_income_changes == {'Salary': [Decimal('2500'), Decimal('0')],
                              'Shopping': [Decimal('0'), Decimal('0')]}
    assert quarter_expenses_changes == {'Salary': [Decimal('0'), Decimal('0')],
                              'Shopping': [Decimal('0'), Decimal('50')]}

    month_periods, month_income_changes, month_expenses_changes = categories_changes(tr_balances, Period.MONTH)
    assert month_periods == [datetime.date(2020, 6, 1), datetime.date(2020, 7, 1)]
    assert month_income_changes == {'Salary': [Decimal('2500'), Decimal('0')],
                              'Shopping': [Decimal('0'), Decimal('0')]}
    assert month_expenses_changes == {'Salary': [Decimal('0'), Decimal('0')],
                              'Shopping': [Decimal('0'), Decimal('50')]}

    week_periods, week_income_changes, week_expenses_changes = categories_changes(tr_balances, Period.WEEK)
    assert week_periods == [datetime.date(2020, 6, 22), datetime.date(2020, 6, 29), datetime.date(2020, 7, 6)]
    assert week_income_changes == {'Salary': [Decimal('2500'), Decimal('0'), Decimal('0')],
                              'Shopping': [Decimal('0'), Decimal('0'), Decimal('0')]}
    assert week_expenses_changes == {'Salary': [Decimal('0'), Decimal('0'), Decimal('0')],
                              'Shopping': [Decimal('0'), Decimal('0'), Decimal('50')]}

    day_periods, day_income_changes, day_expenses_changes = categories_changes(tr_balances, Period.DAY)
    assert len(day_periods) == 15
    assert day_periods[0] == datetime.date(2020, 6, 28)
    assert day_periods[-1] == datetime.date(2020, 7, 12)
    assert day_income_changes['Salary'][0] == Decimal('2500')
    assert day_income_changes['Salary'][-1] == Decimal('0')
    assert day_expenses_changes['Shopping'][0] == Decimal('0')
    assert day_expenses_changes['Shopping'][-1] == Decimal('50')

File: tests/test-totals.js
const fs = require("fs");

const INGTransaction = require("../src/js/lib/sources/ing").INGTransaction;
const transactions = require("../src/js/lib/transactions");
const Parameters = require("../src/js/lib/parameters").Parameters;
const totals = require("../src/js/lib/totals");
const DECIMAL = require("../src/js/lib/utils.js").DECIMAL;

const transactionFilePath = "tests/data/test_transactions.csv";
const parametersFilePath = "tests/data/test_parameters.json";

/**
 * @return {Parameters} Parameters that can be used for testing.
 */
function getTestParameters() {
  return Parameters.import(fs.readFileSync(parametersFilePath).toString());
}
/**
 * @return {Array<Transaction>} Transactions that can be used for testing.
 */
function getTestTransactions() {
  const transactionData = fs.readFileSync(transactionFilePath).toString();
  const uncategorizedTransactions = INGTransaction.loadTransactions(
    transactionData
  ).map((tr) => tr.toTransaction());

  return uncategorizedTransactions;
}

describe("Totals", function () {
  const testParameters = getTestParameters();
  const testTransactions = getTestTransactions();
  const testTrBalances = transactions.transactionBalances(
    testTransactions.map((tr) => tr.categorize(testParameters.categories)),
    testParameters.accounts
  );

  it("can determine categories", function () {
    expect(totals.getCategories(testTrBalances)).toEqual([
      "Salary",
      "Shopping",
    ]);
  });

  it("can generate a year-long range up to the last transaction", function () {
    expect(totals.lastYearRange(testTrBalances)).toEqual([
      new Date("2019-07-12"),
      new Date("2020-07-12"),
    ]);
  });

  it("can floor a date to the first day of a period", function () {
    expect(
      totals.floorPeriod(new Date("2020-05-20"), totals.Period.YEAR)
    ).toEqual(new Date("2020-01-01"));
    expect(
      totals.floorPeriod(new Date("2200-12-31"), totals.Period.YEAR)
    ).toEqual(new Date("2200-01-01"));
    expect(
      totals.floorPeriod(new Date("2000-01-01"), totals.Period.YEAR)
    ).toEqual(new Date("2000-01-01"));

    expect(
      totals.floorPeriod(new Date("2020-05-20"), totals.Period.QUARTER)
    ).toEqual(new Date("2020-04-01"));
    expect(
      totals.floorPeriod(new Date("2200-12-31"), totals.Period.QUARTER)
    ).toEqual(new Date("2200-10-01"));
    expect(
      totals.floorPeriod(new Date("2000-01-01"), totals.Period.QUARTER)
    ).toEqual(new Date("2000-01-01"));

    expect(
      totals.floorPeriod(new Date("2020-05-20"), totals.Period.MONTH)
    ).toEqual(new Date("2020-05-01"));
    expect(
      totals.floorPeriod(new Date("2200-12-31"), totals.Period.MONTH)
    ).toEqual(new Date("2200-12-01"));
    expect(
      totals.floorPeriod(new Date("2000-01-01"), totals.Period.MONTH)
    ).toEqual(new Date("2000-01-01"));

    expect(
      totals.floorPeriod(new Date("2020-05-20"), totals.Period.WEEK)
    ).toEqual(new Date("2020-05-18"));
    expect(
      totals.floorPeriod(new Date("2200-12-31"), totals.Period.WEEK)
    ).toEqual(new Date("2200-12-29"));
    expect(
      totals.floorPeriod(new Date("2000-01-01"), totals.Period.WEEK)
    ).toEqual(new Date("1999-12-27"));

    expect(
      totals.floorPeriod(new Date("2020-05-20"), totals.Period.DAY)
    ).toEqual(new Date("2020-05-20"));
    expect(
      totals.floorPeriod(new Date("2200-12-31"), totals.Period.DAY)
    ).toEqual(new Date("2200-12-31"));
    expect(
      totals.floorPeriod(new Date("2000-01-01"), totals.Period.DAY)
    ).toEqual(new Date("2000-01-01"));
  });

  it("can calculate the first day of the next period", function () {
    expect(
      totals.nextPeriod(new Date("2020-05-20"), totals.Period.YEAR)
    ).toEqual(new Date("2021-01-01"));
    expect(
      totals.nextPeriod(new Date("2200-12-31"), totals.Period.YEAR)
    ).toEqual(new Date("2201-01-01"));
    expect(
      totals.nextPeriod(new Date("2000-01-01"), totals.Period.YEAR)
    ).toEqual(new Date("2001-01-01"));

    expect(
      totals.nextPeriod(new Date("2020-05-20"), totals.Period.QUARTER)
    ).toEqual(new Date("2020-07-01"));
    expect(
      totals.nextPeriod(new Date("2200-12-31"), totals.Period.QUARTER)
    ).toEqual(new Date("2201-01-01"));
    expect(
      totals.nextPeriod(new Date("2000-01-01"), totals.Period.QUARTER)
    ).toEqual(new Date("2000-04-01"));

    expect(
      totals.nextPeriod(new Date("2020-05-20"), totals.Period.MONTH)
    ).toEqual(new Date("2020-06-01"));
    expect(
      totals.nextPeriod(new Date("2200-12-31"), totals.Period.MONTH)
    ).toEqual(new Date("2201-01-01"));
    expect(
      totals.nextPeriod(new Date("2000-01-01"), totals.Period.MONTH)
    ).toEqual(new Date("2000-02-01"));

    expect(
      totals.nextPeriod(new Date("2020-05-20"), totals.Period.WEEK)
    ).toEqual(new Date("2020-05-25"));
    expect(
      totals.nextPeriod(new Date("2200-12-31"), totals.Period.WEEK)
    ).toEqual(new Date("2201-01-05"));
    expect(
      totals.nextPeriod(new Date("2000-01-01"), totals.Period.WEEK)
    ).toEqual(new Date("2000-01-03"));

    expect(
      totals.nextPeriod(new Date("2020-05-20"), totals.Period.DAY)
    ).toEqual(new Date("2020-05-21"));
    expect(
      totals.nextPeriod(new Date("2200-12-31"), totals.Period.DAY)
    ).toEqual(new Date("2201-01-01"));
    expect(
      totals.nextPeriod(new Date("2000-01-01"), totals.Period.DAY)
    ).toEqual(new Date("2000-01-02"));
  });

  it("can calculate the dates 1/3rd and 2/3rds through a period", function () {
    expect(
      totals.periodThirds(new Date("2020-05-01"), totals.Period.YEAR)
    ).toEqual([new Date("2020-05-02"), new Date("2020-09-01")]);
    expect(
      totals.periodThirds(new Date("2020-05-01"), totals.Period.QUARTER)
    ).toEqual([new Date("2020-05-01"), new Date("2020-05-31")]);
    expect(
      totals.periodThirds(new Date("2020-05-01"), totals.Period.MONTH)
    ).toEqual([new Date("2020-05-11"), new Date("2020-05-21")]);
    expect(
      totals.periodThirds(new Date("2020-05-01"), totals.Period.WEEK)
    ).toEqual([new Date("2020-04-29"), new Date("2020-05-01")]);
    expect(
      totals.periodThirds(new Date("2020-05-01"), totals.Period.DAY)
    ).toEqual([new Date("2020-05-01 08:00Z"), new Date("2020-05-01 16:00Z")]);
  });

  it("can calculate the dates half through a period", function () {
    expect(
      totals.periodHalves(new Date("2020-05-01"), totals.Period.YEAR)
    ).toEqual(new Date("2020-07-02"));
    expect(
      totals.periodHalves(new Date("2020-05-01"), totals.Period.QUARTER)
    ).toEqual(new Date("2020-05-16"));
    expect(
      totals.periodHalves(new Date("2020-05-01"), totals.Period.MONTH)
    ).toEqual(new Date("2020-05-16"));
    expect(
      totals.periodHalves(new Date("2020-05-01"), totals.Period.WEEK)
    ).toEqual(new Date("2020-04-30"));
    expect(
      totals.periodHalves(new Date("2020-05-01"), totals.Period.DAY)
    ).toEqual(new Date("2020-05-01 12:00Z"));
  });

  it("can calculate the changes per period grouped by category", function () {
    const [
      yearPeriods,
      yearIncomeChanges,
      yearExpensesChanges,
    ] = totals.categoriesChanges(testTrBalances, totals.Period.YEAR);
    expect(yearPeriods).toEqual([new Date("2020-01-01")]);
    expect(yearIncomeChanges).toEqual({
      Salary: [2500 * DECIMAL],
      Shopping: [0],
    });
    expect(yearExpensesChanges).toEqual({
      Salary: [0],
      Shopping: [50 * DECIMAL],
    });

    const [
      quarterPeriods,
      quarterIncomeChanges,
      quarterExpensesChanges,
    ] = totals.categoriesChanges(testTrBalances, totals.Period.QUARTER);
    expect(quarterPeriods).toEqual([
      new Date("2020-04-01"),
      new Date("2020-07-01"),
    ]);
    expect(quarterIncomeChanges).toEqual({
      Salary: [2500 * DECIMAL, 0],
      Shopping: [0, 0],
    });
    expect(quarterExpensesChanges).toEqual({
      Salary: [0, 0],
      Shopping: [0, 50 * DECIMAL],
    });

    const [
      monthPeriods,
      monthIncomeChanges,
      monthExpensesChanges,
    ] = totals.categoriesChanges(testTrBalances, totals.Period.MONTH);
    expect(monthPeriods).toEqual([
      new Date("2020-06-01"),
      new Date("2020-07-01"),
    ]);
    expect(monthIncomeChanges).toEqual({
      Salary: [2500 * DECIMAL, 0],
      Shopping: [0, 0],
    });
    expect(monthExpensesChanges).toEqual({
      Salary: [0, 0],
      Shopping: [0, 50 * DECIMAL],
    });

    const [
      weekPeriods,
      weekIncomeChanges,
      weekExpensesChanges,
    ] = totals.categoriesChanges(testTrBalances, totals.Period.WEEK);
    expect(weekPeriods).toEqual([
      new Date("2020-06-22"),
      new Date("2020-06-29"),
      new Date("2020-07-06"),
    ]);
    expect(weekIncomeChanges).toEqual({
      Salary: [2500 * DECIMAL, 0, 0],
      Shopping: [0, 0, 0],
    });
    expect(weekExpensesChanges).toEqual({
      Salary: [0, 0, 0],
      Shopping: [0, 0, 50 * DECIMAL],
    });

    const [
      dayPeriods,
      dayIncomeChanges,
      dayExpensesChanges,
    ] = totals.categoriesChanges(testTrBalances, totals.Period.DAY);
    expect(dayPeriods.length).toEqual(15);
    expect(dayPeriods[0]).toEqual(new Date("2020-06-28"));
    expect(dayPeriods[dayPeriods.length - 1]).toEqual(new Date("2020-07-12"));
    expect(dayIncomeChanges["Salary"][0]).toEqual(2500 * DECIMAL);
    expect(dayIncomeChanges["Salary"].slice(-1)[0]).toEqual(0);
    expect(dayExpensesChanges["Shopping"][0]).toEqual(0);
    expect(dayExpensesChanges["Shopping"].slice(-1)[0]).toEqual(50 * DECIMAL);
  });
});
